File: d3d9/RefFrame.cpp
#include "BasicTypes.h"
#include "RefFrame.h"

namespace eve {
	RefFrame::RefFrame(RefFrame* p) : m_type(REF_FRAME) {
		m_localMatrix.setIdentity();
		m_worldMatrix.setIdentity();
		if (p) {
			m_valid = false;
			p->addRef();
			m_parent= p;
			m_child	= 0;
			m_next	= p->m_child;
			p->m_child = this;
		} else {
			m_valid = true;
			m_parent= 0;
			m_child	= 0;
			m_next	= 0;
		}
	}

	RefFrame::~RefFrame() {
		if (m_parent) {
			RefFrame* i;
			for (i = m_parent->m_child; i->m_next != 0; i = i->m_next) {
				if (i->m_next == this) {
					i->m_next = m_next;
					break;
				}
			}
			if (m_parent->m_child == this) {
				m_parent->m_child = m_next;
			}
			m_parent->rmvRef();
		}
	}

	void RefFrame::translate(const Vector1x4& v, const RefFrame* relative2) {
		if (relative2 == this || relative2 == 0) { // relative to own axes
			const f32 x = m_localMatrix.m_41;
			const f32 y = m_localMatrix.m_42;
			const f32 z = m_localMatrix.m_43;
			m_localMatrix.m_41 = x + (v.m_x * m_localMatrix.m_11) + (v.m_y * m_localMatrix.m_21) + (v.m_z * m_localMatrix.m_31);
			m_localMatrix.m_42 = y + (v.m_x * m_localMatrix.m_12) + (v.m_y * m_localMatrix.m_22) + (v.m_z * m_localMatrix.m_32);
			m_localMatrix.m_43 = z + (v.m_x * m_localMatrix.m_13) + (v.m_y * m_localMatrix.m_23) + (v.m_z * m_localMatrix.m_33);

		} else if (relative2->m_parent == 0) { // relative to root axes
			Vector1x4 d = relative2->map(Vector1x4(v.m_x, v.m_y, v.m_z, 0.f), m_parent);
			m_localMatrix.m_41 += d.m_x;
			m_localMatrix.m_42 += d.m_y;
			m_localMatrix.m_43 += d.m_z;

		} else if (relative2 == m_parent) { // relative to parent axes
			m_localMatrix.m_41 += v.m_x;
			m_localMatrix.m_42 += v.m_y;
			m_localMatrix.m_43 += v.m_z;

		} else { // relative to arbitrary axes
			EVE_ASSERT(this->m_parent);
			Vector1x4 d = relative2->map(Vector1x4(v.m_x, v.m_y, v.m_z, 0.f), m_parent);
			m_localMatrix.m_41 += d.m_x;
			m_localMatrix.m_42 += d.m_y;
			m_localMatrix.m_43 += d.m_z;
		}
		invalidateSubtree();
	}

	void RefFrame::rotateX(f32 theta, const RefFrame* relative2) {
		if (relative2 == this || relative2 == 0) { // relative to own axes
			m_localMatrix = Matrix4x4(Matrix4x4::ROTATE_X, theta) * m_localMatrix;

		} else if (relative2 == m_parent) { // relative to parent
			m_localMatrix = m_localMatrix * Matrix4x4(Matrix4x4::ROTATE_X, theta);

		} else { // relative to arbitrary axes
			EVE_ASSERT(false);
		}
		invalidateSubtree();
	}
	void RefFrame::rotateY(f32 theta, const RefFrame* relative2) {	
		if (relative2 == this || relative2 == 0) { // relative to own axes
			m_localMatrix = Matrix4x4(Matrix4x4::ROTATE_Y, theta) * m_localMatrix;

		} else if (relative2 == m_parent) { // relative to parent
			m_localMatrix = m_localMatrix * Matrix4x4(Matrix4x4::ROTATE_Y, theta);

		} else {
			EVE_ASSERT(false);
		}
		invalidateSubtree();
	}
	void RefFrame::rotateZ(f32 theta, const RefFrame* relative2) {
		if (relative2 == this || relative2 == 0) { // relative to own axes
			m_localMatrix = Matrix4x4(Matrix4x4::ROTATE_Z, theta) * m_localMatrix;

		} else if (relative2 == m_parent) { // relative to parent
			m_localMatrix = m_localMatrix * Matrix4x4(Matrix4x4::ROTATE_Z, theta);

		} else {
			EVE_ASSERT(false);
		}
		invalidateSubtree();
	}
	Vector1x4 RefFrame::map(const Vector1x4& v, RefFrame* tgt) const {
		if (tgt == m_parent) { 
			return v * m_localMatrix;
		} else if (tgt->m_parent == 0) { 
			return v * getWorldMatrix();
		} else {
			return v * getWorldMatrix() * !tgt->getWorldMatrix();
		}
	}

	Vector1x4 RefFrame::getPos(const RefFrame* tgt) const {
		static const Vector1x4 origin(0.0f, 0.0f, 0.0f, 1.0f);
		if (tgt == m_parent) {
			return Vector1x4(	m_localMatrix.m_41, 
								m_localMatrix.m_42, 
								m_localMatrix.m_43, 1.f);
		} else if (tgt->m_parent == 0) {
			return origin * getWorldMatrix();
		} else {
			return origin * getWorldMatrix() * !tgt->getWorldMatrix();
		}
	}
	const Matrix4x4& RefFrame::getLocalMatrix() const {
		return m_localMatrix;
	}
	const Matrix4x4& RefFrame::getWorldMatrix() const {
		validateAscending();
		return m_worldMatrix;
	}

	void RefFrame::setLocalMatrix(const Matrix4x4& mtx) {
		m_localMatrix = mtx;
		invalidateSubtree();
	}
	void RefFrame::setWorldMatrix(const Matrix4x4& mtx) {
		// worldMatrix = localMatrix * parent->worldMatrix
		// worldMatrix / parent->worldMatrix = localMatrix
		m_localMatrix = mtx * !m_parent->getWorldMatrix();
		invalidateSubtree();
	}

	void RefFrame::invalidateSubtree() {
		if (m_valid) {
			for (RefFrame* i = m_child; i != 0; i = i->m_next) {
				i->invalidateSubtree();
			}
			m_valid = false;
			invalidateSubClass();
		}
	}	
	void RefFrame::validateAscending() const {
		if (!m_valid) {
			if (m_parent) {
				m_parent->validateAscending();
				m_worldMatrix = m_localMatrix * m_parent->m_worldMatrix;
			}
			m_valid = true;
		}
	}
}

File: webgl/src/RefFrame.js
import Matrix4x4, {_11, _12, _13, _14, _21, _22, _23, _24, _31, _32, _33, _34, _41, _42, _43, _44 } from './Matrix4x4.js';
import Vector1x4 from './Vector1x4.js';

export default class RefFrame {
    constructor(parent = null, node = null) {
        this.validSubtree = true;
        this.parent       = null;
        this.child        = null;
        this.next         = null;
        this.localM       = new Matrix4x4();
        this.modelM       = new Matrix4x4();

        if (parent instanceof RefFrame) {
            this.validSubtree  = false;
            this.parent        = parent;
            this.next          = parent.child;
            this.parent.child  = this;
        }

        if (node) {
           this.modelMatrix = new Matrix4x4(node.modelMatrix); 
        }
    }

    invalidateSubtree() {
        if (this.validSubtree) {
            for (let i = this.child; i !== null; i = i.next) {
                i.invalidateSubtree();
            }
            this.validSubtree = false;
        }
    }

    validateAscending() {
        if (!this.validSubtree) {
            if (this.parent) {
                this.parent.validateAscending();
                this.modelM = this.localM.mul(this.parent.modelM);
            }
            this.validSubtree = true;
        }
    }

    *children() {
        for (let sibling = this.child; sibling !== null; sibling = sibling.next) { yield sibling; }
    }

    get localMatrix() {
        return this.localM;
    }

    set localMatrix(matrix) {
        this.localM = new Matrix4x4(matrix);
        this.invalidateSubtree();
    }

    get modelMatrix() {
        this.validateAscending();
        return this.modelM;
    }

    set modelMatrix(matrix) {
        // worldMatrix = localMatrix * parent->worldMatrix
        // worldMatrix / parent->worldMatrix = localMatrix
        if (this.parent instanceof RefFrame) {
            this.localM = matrix.mul(this.parent.modelMatrix.inverse());
            this.invalidateSubtree();
        }
    }

    translate(v, relative2) {
        if (relative2 === this || !relative2) { // relative to own axes
            const x = this.localM.m[_41];
            const y = this.localM.m[_42];
            const z = this.localM.m[_43];
            this.localM.m[_41] = x + (v.x * this.localM.m[_11]) + (v.y * this.localM.m[_21]) + (v.z * this.localM.m[_31]);
            this.localM.m[_42] = y + (v.x * this.localM.m[_12]) + (v.y * this.localM.m[_22]) + (v.z * this.localM.m[_32]);
            this.localM.m[_43] = z + (v.x * this.localM.m[_13]) + (v.y * this.localM.m[_23]) + (v.z * this.localM.m[_33]);

        } else if (relative2.parent === null) { // relative to root axes
            const d = relative2.mapPos(new Vector1x4(v.x, v.y, v.z, 0.0), this.parent);
            this.localM.m[_41] += d.m_x;
            this.localM.m[_42] += d.m_y;
            this.localM.m[_43] += d.m_z;

        } else if (relative2 === this.parent) { // relative to parent axes
            this.localM.m[_41] += v.x;
            this.localM.m[_42] += v.y;
            this.localM.m[_43] += v.z;

        } else { // relative to arbitrary axes
            const d = relative2.mapPos(new Vector1x4(v.x, v.y, v.z, 0.0), this.parent);
            this.localM.m[_41] += d.m_x;
            this.localM.m[_42] += d.m_y;
            this.localM.m[_43] += d.m_z;
        }
        this.invalidateSubtree();
    }

    rotateX(theta, relative2) {
        if (relative2 === this || !relative2) { // relative to own axes
            const rotx = Matrix4x4.createRx(theta);
            this.localMatrix = rotx.mul(this.localMatrix);
        } else if (relative2 === this.parent) { // relative to parent
            const rotx = Matrix4x4.createRx(theta);
            this.localMatrix = this.localMatrix.mul(rotx);
        } else { // relative to arbitrary axes
            throw new Error('Cannot rotate relative to arbitrary axis!');
        }
        this.invalidateSubtree();
    }

    rotateY(theta, relative2) {
        if (relative2 === this || !relative2) { // relative to own axes
            const roty = Matrix4x4.createRy(theta);
            this.localMatrix = roty.mul(this.localMatrix);
        } else if (relative2 === this.parent) { // relative to parent
            const roty = Matrix4x4.createRy(theta);
            this.localMatrix = this.localMatrix.mul(roty);
        } else { // relative to arbitrary axes
            throw new Error('Cannot rotate relative to arbitrary axis!');
        }
        this.invalidateSubtree();
    }

    rotateZ(theta, relative2) {
        if (relative2 === this || !relative2) { // relative to own axes
            const rotz = Matrix4x4.createRz(theta);
            this.localMatrix = rotz.mul(this.localMatrix);
        } else if (relative2 === this.parent) { // relative to parent
            const rotz = Matrix4x4.createRz(theta);
            this.localMatrix = this.localMatrix.mul(rotz);
        } else { // relative to arbitrary axes
            throw new Error('Cannot rotate relative to arbitrary axis!');
        }
        this.invalidateSubtree();
    }

    mapPos(v, tgt) {
        if (tgt === this.parent) {
            return v.mul(this.localMatrix);
        } else if (tgt.parent === null) {
            return v.mul(this.modelMatrix);
        } else {
            return v.mul(this.modelMatrix).mul(tgt.modelMatrix.inverse());
        }
    }
}

